### PR TITLE
Fix for issue 5575 along with testing

### PR DESCRIPTION
**Pull request by KyleBsingh**
This is a pull request to resolve issue #5575.

Tracing calls made in plot_date() I realized it was ignoring timezones as a result of the axis being set while plotting, then trying to change the time zone.

In line 1427 of axis.py we have `converter = munits.registry.get_converter(data)`. Unfortunately for setting the timezone the converter will be the same as when it was plotted originally. Thus in the following lines 1433-1435 we will not be able to set_units.

```python
default = self.converter.default_units(data, self)
if default is not None and self.units is None:
    self.set_units(default)
```

My solution to this was to make sure that instead of the following in _axes.py's plot_data:
```python
if xdate:
    self.xaxis_date(tz)
if ydate:
    self.yaxis_date(tz)

ret = self.plot(x, y, fmt, **kwargs)
```

we have:
```python
ret = self.plot(x, y, fmt, **kwargs)
if xdate:
    self.xaxis_date(tz)
if ydate:
    self.yaxis_date(tz)
```

This ensures the plot has the correct timezone, and it also plots the correct data onto the plot.
I have added test cases to /lib/matplotlib/tests/test_axes which has tests for the same and different timezones in the x axis, y axis, and both the x and y axis

### CODE REVIEW

**Comment by tacaswell** [comment]
Can you add only png tests for this (and re-base the pdf/svg tests out of existence)?  The reason is to keep the size of the repo down and to keep the tests as fast as possible.  There is nothing backend dependent about this so testing it only on Agg should be enough.

Other wise :+1: 

It is great when problems have simple fixes!

Can you also put the explanation of why this fixed the problem in the commit message?

**Comment by KyleBsingh** [comment]
Done. Let me know if I missed anything in the pull request/commit.

**Comment by tacaswell** [comment]
Can you rebase this to only one commit?  Currently the svg/pdf are still in the history which means they are (forever) in the repository.

**Comment by KyleBsingh** [comment]
Done!

**Comment by tacaswell** [comment]
If you just keep pu

**Comment by tacaswell** [comment]
If you just keep (force) pushing to this branch github will 'do the right thing'.  PRs track branches, not commits.

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -4335,6 +4337,50 @@ def test_pandas_indexing_hist():
     fig, axes = plt.subplots()
     axes.hist(ser_2)
 
+@image_comparison(baseline_images=['date_timezone_x'],
+                  extensions=['png'])
+def test_date_timezone_x():
+    #Tests issue 5575
+    time_index = [pytz.timezone('Canada/Eastern').localize(datetime.datetime(year=2016, month=2, day=22, hour=x)) for x in range(3)]
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: This line is too long; it should be wrapped at 79 characters.

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -4335,6 +4337,50 @@ def test_pandas_indexing_hist():
...
+
+    #Same Timezone
+    fig = plt.figure(figsize=(20,12))
+    plt.subplot(2, 1, 1)
+    plt.plot_date(time_index, [3]*3, tz='Canada/Eastern')
+
+    #Different Timezone
+    plt.subplot(2, 1, 2)
+    plt.plot_date(time_index, [3]*3, tz='UTC')
+
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Please use two empty lines between functions (this file is not yet checked against PEP8.)

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -4335,6 +4337,50 @@ def test_pandas_indexing_hist():
     fig, axes = plt.subplots()
     axes.hist(ser_2)
 
+@image_comparison(baseline_images=['date_timezone_x'],
+                  extensions=['png'])
+def test_date_timezone_x():
+    #Tests issue 5575
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Missing space after `#` (and in all other functions.)

File: lib/matplotlib/tests/test_axes.py [hunk 4]
```diff
@@ -4335,6 +4337,50 @@ def test_pandas_indexing_hist():
...
+    time_index = [pytz.timezone('Canada/Eastern').localize(datetime.datetime(year=2016, month=2, day=22, hour=x)) for x in range(3)]
+
+    #Same Timezone
+    fig = plt.figure(figsize=(20,12))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Missing space after `,`.

File: lib/matplotlib/tests/test_axes.py [hunk 5]
```diff
@@ -4335,6 +4337,50 @@ def test_pandas_indexing_hist():
...
+    plt.subplot(2, 1, 1)
+    plt.plot_date(time_index, [3]*3, tz='Canada/Eastern')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Missing spaces around `*`.

**Comment by KyleBsingh** [comment]
Sorry about earlier, I mis-clicked "Close and comment" instead of "comment". I have rebased and made the pep8 changes that I admittedly forgot to do. Thank you

**Comment by tacaswell** [comment]
merged locally to resolve minor conflict ee1a98cfa9d0092ec3a4f8f85972315c9eaee600

backported to v1.5.x as 44afc73

**Pull request merged**